ethereum/core/build.gradle

```diff
@@ -38,6 +38,7 @@ sourceSets {
 }
 
 dependencies {
+  implementation 'info.picocli:picocli:4.6.1'
```

Comment:
Can we avoid that somehow?

Reply:
Sadly, no. This was necessary since putting the configuration classes in besu/main would cause a circular dependency between that and ethereum/core. I think this is important for moving towards a package-by-feature approach, which we will need more and more of in the future as we divide up the codebase..

Reply:
Please expand a lot more - we’re dividing up the code base?

Reply:
When did we go to a package by feature approach?  Not that I'm against it, it's just that it seems like the sort of thing that should be discussed among more than just ConsenSys maintainer.

That being said, the current dogma would be to put this feature in a stand alone config object (or add it to an existing config object) and have the BesuCommand handle marshaling the features in and out.  Wiring in the PicoCLI flags binds the library code to a CLI style and makes it less suitable to library inclusion. It's mixing concerns and strongly tying us to one CLI library.

Reply:
At a high level, my thinking is that CLI configuration should be fully severable from the operation and packaging of the code.

Reply:
Yes, that is my thinking too @shemnon , and seems to be in the spirit of the PicoCLI design. It naturally allows you to avoid these circular dependencies. I tried to implement this via a standalone config object and ended up writing more code than was valuable, to move around a single config option. Putting that config class in between the option and the CodeCache was what caused the circular dependency.

@atoulme I was referring to the last contributors call where we discussed separating out the EVM into it's own library. Between that and other discussions around providing MEV functionalities, it seems like we are heading in the direction of trying to modularize Besu. 

If there exists any documentation on why Besu is divided into the gradle projects and java packages that it is, I am more than happy to study it.

Reply:
@jflo there is a leap between saying we want to carve out the EVM so we can make it a library, and saying we're modularizing Besu, or applying what you call a package by feature approach.
Please do document your approach and submit it to contributors so we can all review.
* What are the modules?
* How is configuration and dependency injection working?
* What is the path to executables? - for example, you add picocli as a dependency right in here, but frankly, that should be decoupled away into a separate config module so you can have an ethereum lib.

Reply:
A great deal of Besu is already packaged by feature. For instance, looking at the besu/ethereum/eth module, in the org.hyperledger.besu.ethereum.eth.sync package, I see many different classes with different functions, all serving the synchronization feature. I'm not sure why this seems like a leap? Is it the lack of an intermediary Configuration class between the Option and the class using it?

Reply:
> Wiring in the PicoCLI flags binds the library code to a CLI style and makes it less suitable to library inclusion. It's mixing concerns and strongly tying us to one CLI library.

@shemnon can you elaborate on this? The flags are defined in the Options class, just like everywhere else, so a change in the CLI library still only requires updates to a single class. I think the part I'm really struggling with is the duplication of all the fields across the Options and Configuration class. I'm unsure why that is better and what it is buying me for this case.

Reply:
The options class should be in the besu module... `besu/src/main/java/org/hyperledger/besu/cli/options/unstable/`, which should have a `toDomainObject` method exporting the CLI value to a domain object that live in the core module.  Lie what `DnsOptions` does (a fairly simple example).

Reply:
Ok, i did this, but only up to a point. All of the other Configurations (such as `DnsOptions`) have a much shorter distance down into the codebase to travel than this one does. Most are used in the controller or service layers, while the contract code cache is used from multiple implementations of WorldState or WorldView. Doing this the exact same way would add the Configuration class to the signatures of tons of code. (I went down this path and have a branch somewhere if you would like to see the horrors). I thinks this singleton approach is a less bad compromise that 

1. keeps the Option as the single place that knows about PicoCli
2. avoids making dozens of other classes aware of a Configuration they don't need, and only pass down to lower layers.